File: solutions/csharp/need-for-speed/1/NeedForSpeed.cs
class RemoteControlCar
{
    private int speed;
    private int drain;
    private int driven = 0;
    private int battery = 100;

    public int getSpeed() => this.speed;
    public int getDrain() => this.drain;
    
    public RemoteControlCar(int speed, int drain) {
        this.speed = speed;
        this.drain = drain;
    }

    public bool BatteryDrained() => battery < drain ? true : false;

    public int DistanceDriven() => driven;

    public void Drive()
    {
        if (!BatteryDrained()) {
            battery -= drain;
            driven += speed;
        }
    }

    public static RemoteControlCar Nitro() => new RemoteControlCar(50, 4);
}

class RaceTrack
{
    private int distance;
    
    public RaceTrack(int distance) => this.distance = distance;

    public bool TryFinishTrack(RemoteControlCar car) => Convert.ToDouble(distance) / car.getSpeed() * car.getDrain() <= 100 ? true : false;
}

File: solutions/java/need-for-speed/1/src/main/java/NeedForSpeed.java
class NeedForSpeed {
    private int driven = 0;
    private int battery = 100;
    private int speed;
    private int batteryDrain;
    
    NeedForSpeed(int speed, int batteryDrain) {
        this.speed = speed;
        this.batteryDrain = batteryDrain;
    }

    public boolean batteryDrained() {
        return battery < batteryDrain;
    }

    public int distanceDriven() {
        return driven;
    }

    public void drive() {
        if (!batteryDrained()) {
            driven += speed;
            battery -= batteryDrain;
        }
    }

    public static NeedForSpeed nitro() {
        return new NeedForSpeed(50, 4);
    }
}

class RaceTrack {
    private int distance;
    
    RaceTrack(int distance) {
        this.distance = distance;
    }

    public boolean canFinishRace(NeedForSpeed car) {
        while (!car.batteryDrained() && car.distanceDriven() < distance) {
            car.drive();
        }

        return car.distanceDriven() >= distance;
    }
}
